### PR TITLE
Allow redirection of lcf console messages

### PR DESCRIPTION
**Pull request by Ghabry**
First commit: The chunk dumping during Skip didnt really work except on Windows. I finally fixed it (useful for Maniacs Patch to find unknown chunks). Also is now always on, a missing chunk is now always a bug because we know about all common ones.

Second commit: **I want your feedback here**. Before I implement this completely (is some work) please tell me if the general approach is fine for you.
I want to allow custom output handlers for lcf diagnostics. E.g. for the Player they could be written to easyrpg_log.txt. Currently they are completely lost which is bad for handling bug reports. As vital information is missing.

To redirect it library users can invoke lcf::LogHandler::SetHandler(_Handlerfunctionhere_)

### CODE REVIEW

**Comment by fdelapena** [comment]
There's a good chance to add more levels, such `trace`, `info`, `notice`, while `critical`, `alert`, `emergency` feel a bit unneeded.
At least those for matching not just with Player, but cover at least the levels existing in JavaScript's `console.xxx()`. Emscripten shell currently uses `console.error()` for everything but this is easy to fix and helps a lot to filter details from the browser developer tools.

**Comment by aleck099** [comment]
Why is exceptions disabled in liblcf?

**Comment by Ghabry** [comment]
Same reason why we disable RTTI:

Portability. We have some quite low spec homebrew platforms that we support.
They are low on memory and load the executable into memory so every kilobyte counts.
Also exceptions are also not well supported in emscripten.

We also do not want to maintain two codepaths. So no exception it is.

**Comment by aleck099** [comment]
But is printing to stderr a good choice?

**Comment by Ghabry** [comment]
Uhm no that is why I plan to make this configurable in this PR?

Currently errors land on stderr.
With this PR consumers of this library (e.g. the Player) can hook it into their logging system (which logs into easyrpg_log.txt)

**Comment by aleck099** [comment]
Looking forward to begin merged

**Comment by Ghabry** [comment]
I rechecked the code now: There is literally nothing finished here yet. Except for one example usage. Can't finish this for 0.8.0 in such short time.

Hot candidate for 0.8.1

**Comment by carstene1ns** [comment]
I think on liblcf side is not much missing anymore. The handler function concept is well-known and works. Usually you get some `void *userdata` context to pass, which we maybe should make available for custom loggers.

**Comment by Ghabry** [comment]
Suitable logging function for use in Player:

```cpp
lcf::LogHandler::SetHandler([](lcf::LogHandler::Level level, StringView message, lcf::LogHandler::UserData) {
    Output::Debug("lcf ({}): {}", lcf::LogHandler::kLevelTags.tag(level), message);
});
```

So currently only written to terminal/logfile.

Not sure if any of the stuff should be reported through the graphical "console" inside the window. Most of that stuff is recoverable (similar to the libpng warnings spamming the shell) and everything critical is already shown as an "Output::Error".

File: src/lcf/log_handler.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/*
+ * This file is part of liblcf. Copyright (c) 2020 liblcf authors.
```
Comment (carstene1ns) [inline src/lcf/log_handler.h:2]: copyright years here and in other `log*` files

**Review by carstene1ns — APPROVED** [review]
I find it a bit messy that at first every log message is made a C-String, then a C++ std::string and last a string_view.
Maybe we should make `fmt` a lcf dependency in the future to hide the conversions.

Apart from the minor nits, LGTM

File: src/log.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+#ifndef LCF_LOG_H
+#define LCF_LOG_H
+
+#include "lcf/log_handler.h"
+
+namespace lcf {
+namespace Log {
+
```
Comment (carstene1ns) [inline src/log.h:17]: One thing that could be added here is argument checking by `g++`
https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-format-function-attribute

I have hidden this behind a preprocessor define here:
```c++
#ifdef __GNUG__
	#define LIKE_PRINTF __attribute__((format(printf, 1, 2)))
#else
	#define LIKE_PRINTF
#endif
```

Comment (Ghabry) [reply src/log.h:23]: done and it even found a bug :D

**Review by fdelapena — APPROVED** [review]

File: src/log.h [hunk 3]
```diff
@@ -12,12 +12,18 @@
...
 
-void Debug(const char* fmt, ...);
-void Warning(const char* fmt, ...);
-void Error(const char* fmt, ...);
+void Debug(const char* fmt, ...) LIKE_PRINTF;
+void Warning(const char* fmt, ...) LIKE_PRINTF;
+void Error(const char* fmt, ...) LIKE_PRINTF;
 
 } // namespace Log
 } // namespace lcf
```
Comment (carstene1ns) [inline src/log.h:29]: Minor nit: we should `#undef LIKE_PRINTF` here to not pollute other projects with internal stuff.

**Review by carstene1ns — APPROVED** [review]

File: src/reader_lcf.cpp [hunk 4]
```diff
@@ -276,21 +279,27 @@ int LcfReader::Peek() {
...
+	ss << std::hex;
+
 	for (uint32_t i = 0; i < chunk_info.length; ++i) {
 		uint8_t byte;
 		LcfReader::Read(byte);
-		fprintf(stderr, "%02X ", byte);
+		ss << std::setfill('0') << std::setw(2) << (int)byte << " ";
 		if ((i+1) % 16 == 0) {
-			fprintf(stderr, "\n");
+			Log::Debug(ss.str().c_str());
```
Comment (carstene1ns) [inline src/reader_lcf.cpp:293]: Android (clang) build fails with `-Wformat-security`:

```c++
Log::Debug("%s", ss.str().c_str());
```

**Pull request merged**